File: zig-generator/src/state.zig
const std = @import("std");
const CovPoint = @import("cov_point.zig").CovPoint;

pub const RegionId = u32;

pub const RegKind = enum(u8) {
    Uninitialized,
    Scalar,
    Pointer,
    Reserved
};

pub const RegInfo = struct {
    kind: RegKind = .Uninitialized,
    region: ?RegionId = null,
    offset: u64 = 0,
};

pub const MemRegion = struct {
    id: RegionId,
    size: u64,
    base_reg: u8,
    watermark: u64,
    label: []const u8,
};

pub const GeneratorState = struct {
    pub const num_gprs:    u8 = 31;
    pub const scratch_reg: u8 = 20;
    pub const sig_pointer: u8 = 21;
    pub const scratch_pointer: u8 = 22;

    allocator: std.mem.Allocator,
    seed: u64,
    rng: std.Random.DefaultPrng,
    regions: std.ArrayList(MemRegion),
    regs: [num_gprs]RegInfo,
    last_written: ?u8,
    scratch_region_id: ?RegionId,
    label_counter: u32,
    coverage_hits: std.EnumSet(CovPoint),

    pub fn init(allocator: std.mem.Allocator, seed: u64) GeneratorState {
        var state = GeneratorState {
            .allocator = allocator,
            .seed = seed,
            .rng = std.Random.DefaultPrng.init(seed),
            .regions = .empty,
            .regs = undefined,
            .last_written = null,
            .scratch_region_id = null,
            .label_counter = 0,
            .coverage_hits = .{},
        };

        for (&state.regs) |*reg| reg.* = .{};
        state.regs[scratch_reg].kind = .Reserved;
        state.regs[sig_pointer].kind = .Reserved;
        state.regs[scratch_pointer].kind = .Reserved;
        return state;
    }


    pub fn deinit(self: *GeneratorState) void { self.regions.deinit(self.allocator); }

    pub fn getRng(self: *GeneratorState) std.Random { return self.rng.random(); }
    pub fn getSeed(self: *const GeneratorState) u64 { return self.seed; }
    pub fn kindOf(self: *const GeneratorState, reg: u8) RegKind { return self.regs[reg].kind; }
    pub fn infoOf(self: *const GeneratorState, reg: u8) RegInfo { return self.regs[reg]; }

    pub fn pickInitializedGpr(self: *GeneratorState) ?u8 {
        var candidates: [num_gprs]u8 = undefined;
        var count: usize = 0;

        for (0..num_gprs) |i| {
            const k = self.regs[i].kind;
            if (k == .Scalar or k == .Pointer) {
                candidates[count] = @intCast(i);
                count += 1;
            }
        }
        if (count == 0) return null;

        return candidates[self.rng.random().uintLessThan(usize, count)];
    }

    pub fn pickWritableGpr(self: *GeneratorState) ?u8 {
        var candidates: [num_gprs]u8 = undefined;
        var count: usize = 0;

        for (0..num_gprs) |i| {
            const k = self.regs[i].kind;
            if (k != .Reserved) {
                candidates[count] = @intCast(i);
                count += 1;
            }
        }
        if (count == 0) return null;

        return candidates[self.rng.random().uintLessThan(usize, count)];
    }

    pub fn pickPointerGpr(self: *GeneratorState, min_headroom: u64) ?u8 {
        var candidates: [num_gprs]u8 = undefined;
        var count: usize = 0;

        for (0..num_gprs) |i| {
            const k = self.regs[i].kind;
            if (k != .Pointer) continue;
            const info = self.regs[i];
            const region_id = info.region orelse continue;
            const mem = self.regions.items[region_id];
            const remaining: u64 = mem.size - info.offset;
            if (remaining >= min_headroom) {
                candidates[count] = @intCast(i);
                count += 1;
            }
        }
        if (count == 0) return null;

        return candidates[self.rng.random().uintLessThan(usize, count)];
    }

    pub fn markScalar(self: *GeneratorState, reg: u8) void {
        self.regs[reg] = .{ .kind = .Scalar };
        self.last_written = reg;
    }
    pub fn markPointer(self: *GeneratorState, reg: u8, region: RegionId, offset: u64) void {
        self.regs[reg] = .{ .kind = .Pointer, .region = region, .offset = offset };
    }

    pub fn markReserved(self: *GeneratorState, reg: u8) void { self.regs[reg] = .{ .kind = .Reserved }; }
    pub fn markUninitialized(self: *GeneratorState, reg: u8) void { self.regs[reg] = .{ .kind = .Uninitialized }; }

    pub fn nextLabelId(self: *GeneratorState) u32 {
        defer self.label_counter += 1;
        return self.label_counter;
    }



    pub fn getRegion(self: *const GeneratorState, id: RegionId) MemRegion { return self.regions.items[id]; }
    pub fn addRegion(self: *GeneratorState, size: u64, base_reg: u8, label: []const u8) !RegionId {
        const id: RegionId = @intCast(self.regions.items.len);
        try self.regions.append(self.allocator, .{
            .id = id,
            .size = size,
            .base_reg = base_reg,
            .watermark = 0,
            .label = label,
        });

        return id;
    }

    pub fn advanceWatermark(self: *GeneratorState, id: RegionId, bytes: u64) void { self.regions.items[id].watermark += bytes; }

    pub fn getScratchRegionId(self: *const GeneratorState) ?RegionId { return self.scratch_region_id; }
    pub fn setScratchRegionId(self: *GeneratorState, id: RegionId) void { self.scratch_region_id = id; }

    pub fn getLastWrittenReg(self: *const GeneratorState) ?u8 { return self.last_written; }

    pub fn hitCoverage(self: *GeneratorState, p: CovPoint) void { self.coverage_hits.insert(p); }
    pub fn getCoverageHits(self: *const GeneratorState) std.EnumSet(CovPoint) { return self.coverage_hits; }

};

File: rust-generator/src/state.rs
use enumset::EnumSet;
use rand::{RngExt};
use rand_mt::Mt64;

use crate::coverage::CovPoint;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum RegKind {
    #[default]
    Uninitialized = 0,
    Scalar,
    Pointer,
    Reserved,
}

pub type RegionId = u32;

pub struct MemRegion {
    pub id: RegionId ,
    pub size: u64,
    pub base_reg: u8,
    pub watermark: u64,
    pub label: String,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct RegInfo {
    kind: RegKind,
    region: Option<RegionId>,
    offset: u64,
}

pub struct GeneratorState {
    seed: u64,
    rng: Mt64,
    regions: Vec<MemRegion>,
    regs: [RegInfo; Self::NUM_GPRS as usize],
    last_written: Option<u8>,
    scratch_region_id: Option<RegionId>,
    label_counter: u32,
    coverage_hits: EnumSet<CovPoint>,
}

impl GeneratorState {
    pub const NUM_GPRS: u8 = 31;
    pub const NUM_GPRS_USIZE: usize = Self::NUM_GPRS as usize;
    pub const SCRATCH_REGISTER: u8 = 20;
    pub const SIG_POINTER: u8 = 21;
    pub const SCRATCH_POINTER: u8 = 22;

    pub fn new(seed: u64) -> Self {
        let mut state = Self {
            seed,
            rng: Mt64::new(seed),
            regions: Vec::new(),
            regs: std::array::from_fn(|_| RegInfo::default()),
            last_written: None,
            scratch_region_id: None,
            label_counter: 0,
            coverage_hits: EnumSet::empty(),
        };
        state.regs[Self::SCRATCH_REGISTER as usize].kind = RegKind::Reserved;
        state.regs[Self::SIG_POINTER as usize].kind = RegKind::Reserved;
        state.regs[Self::SCRATCH_POINTER as usize].kind = RegKind::Reserved;

        state
    }

    pub fn kind_of(&self, reg: u8) -> RegKind { self.regs[reg as usize].kind }
    pub fn info_of(&self, reg: u8) -> RegInfo { self.regs[reg as usize] }
    pub fn rng(&mut self) -> &mut Mt64 { &mut self.rng }
    pub fn seed(&self) -> u64 { self.seed }

    pub fn pick_initialized_gpr(&mut self) -> Option<u8> {
        let mut candidates = [0u8; Self::NUM_GPRS_USIZE];
        let mut count: usize = 0;
        for i in 0..Self::NUM_GPRS_USIZE {
            let k = self.regs[i].kind;
            if matches!(k, RegKind::Scalar | RegKind::Pointer) {
                candidates[count] = i as u8;
                count += 1;
            }
        }
        if count == 0 { return None };

        Some(candidates[self.rng.random_range(0..count)])
    }

    pub fn pick_writable_gpr(&mut self) -> Option<u8> {
        let mut candidates = [0u8; Self::NUM_GPRS_USIZE];
        let mut count: usize = 0;
        for i in 0..Self::NUM_GPRS_USIZE {
            let k = self.regs[i].kind;
            if k != RegKind::Reserved {
                candidates[count] = i as u8;
                count += 1;
            }
        }
        if count == 0 { return None; }


        Some(candidates[self.rng.random_range(0..count)])
    }

    pub fn pick_pointer_gpr(&mut self, min_headroom: u64) -> Option<u8> {
        let mut candidates = [0u8; Self::NUM_GPRS_USIZE];
        let mut count: usize = 0;

        for i in 0..Self::NUM_GPRS_USIZE {
            let info = self.regs[i];
            if info.kind != RegKind::Pointer { continue; }
            let Some(region_id) = info.region else { continue; };
            let mem = &self.regions[region_id as usize];
            let remaining: u64 = mem.size - info.offset;

            if remaining >= min_headroom {
                candidates[count] = i as u8;
                count += 1;
            }
        }
        if count == 0 { return None; }

        Some(candidates[self.rng.random_range(0..count)])
    }

    pub fn mark_scalar(&mut self, reg: u8) {
        self.regs[reg as usize] = RegInfo { kind: RegKind::Scalar, ..RegInfo::default() };
        self.last_written = Some(reg);
    }
    pub fn mark_pointer(&mut self, reg: u8, region: RegionId, offset: u64) { self.regs[reg as usize] = RegInfo { kind: RegKind::Pointer, region: Some(region), offset}; }
    pub fn mark_reserved(&mut self, reg: u8) { self.regs[reg as usize] = RegInfo { kind: RegKind::Reserved, ..RegInfo::default() }; }
    pub fn mark_uninitialized(&mut self, reg: u8) { self.regs[reg as usize] = RegInfo { kind: RegKind::Uninitialized, ..RegInfo::default() }; }


    pub fn next_label_id(&mut self) -> u32 {
        let id = self.label_counter;
        self.label_counter += 1;
        id
    }

    pub fn get_region(&self, id: RegionId) -> &MemRegion { &self.regions[id as usize] }
    pub fn add_region(&mut self, size: u64, base_reg: u8, label: String) -> RegionId {
        let id: RegionId = self.regions.len() as RegionId;
        self.regions.push( MemRegion { 
            id,
            size,
            base_reg,
            watermark: 0,
            label
        });

        id
    }
    pub fn advance_watermark(&mut self, id: RegionId, bytes: u64) { self.regions[id as usize].watermark += bytes; }

    pub fn get_scratch_region_id(&self) -> Option<RegionId> { self.scratch_region_id }
    pub fn set_scratch_region_id(&mut self, id: RegionId) { self.scratch_region_id = Some(id); }

    pub fn get_last_written_reg(&self,) -> Option<u8> { self.last_written }

    pub fn hit_coverage(&mut self, p: CovPoint) { self.coverage_hits.insert(p); }
    pub fn get_coverage_hits(&self) -> EnumSet<CovPoint> { self.coverage_hits }
}
